From 591301bdd418ee99e9812f454e9ffcc8847ad02b Mon Sep 17 00:00:00 2001
From: jjm2473 <1129525450@qq.com>
Date: Mon, 9 Oct 2023 11:43:32 +0800
Subject: [PATCH] PCI: rockchip: dw: fix pcie crash

kernel crash on run `lspci` or load some kmods
```
[   10.474830] Unable to handle kernel access to user memory outside uaccess routines at virtual address 0000000000000020
[   10.478257] Mem abort info:
[   10.478512]   ESR = 0x96000005
[   10.478789]   EC = 0x25: DABT (current EL), IL = 32 bits
[   10.479257]   SET = 0, FnV = 0
[   10.479532]   EA = 0, S1PTW = 0
[   10.479808] Data abort info:
[   10.480061]   ISV = 0, ISS = 0x00000005
[   10.480404]   CM = 0, WnR = 0
[   10.480670] user pgtable: 4k pages, 39-bit VAs, pgdp=0000000005801000
[   10.481240] [0000000000000020] pgd=0000000006839003, p4d=0000000006839003, pud=0000000006839003, pmd=0000000000000000
[   10.482178] Internal error: Oops: 96000005 [#1] SMP
[   10.482613] Modules linked in: r8125(+) igc e1000 e100 ifb dummy nat46 l2tp_ip6 l2tp_ip l2tp_eth ip6_vti ip_vti sit sctp mdio l2tp_netlink l2tp_core ipcomp6 xfrm6_tunnel esp6 ah6 xfrm4_tunnel ipcomp esp4 ah4 ipip ip6_tunnel netlink_diag tunnel6 tunnel4 ip_tunnel udp_diag tcp_diag raw_diag inet_diag hfcpci hfcmulti rpcsec_gss_krb5 dm_raid veth tun snd_pcm_oss snd_mixer_oss snd_hwdep snd_compress nbd mISDN_dsp l1oip mISDN_core xfrm_ipcomp af_key udf minix hfsplus hfs cramfs cifs binfmt_misc br2684 atm aoe multipath bcma raid456 async_raid6_recov async_pq async_xor async_memcpy async_tx raid10 raid1 raid0 linear md_mod nls_iso8859_1 zram zsmalloc natsemi vxlan udp_tunnel ip6_udp_tunnel macsec seqiv echainiv deflate cts authenc crypto_acompress ohci_pci uhci_hcd sata_via sata_sil24 sata_sil ata_piix pata_pdc202xx_old sata_nv pata_artop fsl_mph_dr_of ehci_platform ehci_fsl igb gpio_button_hotplug xfs reiserfs jfs f2fs btrfs xor zstd_compress zlib_deflate xor_neon raid6_pq lzo_compress dm_mirror
[   10.482814]  dm_region_hash dm_log dm_crypt button_hotplug tg3 ssb ptp pps_core encrypted_keys trusted tpm
[   10.491256] CPU: 0 PID: 2400 Comm: kmodloader Not tainted 5.10.176 #0
[   10.491826] Hardware name: FastRhino R68s (DT)
[   10.492219] pstate: a0400085 (NzCv daIf +PAN -UAO -TCO BTYPE=--)
[   10.492752] pc : dw_pcie_link_up+0x10/0x40
[   10.493119] lr : dw_pcie_other_conf_map_bus+0x34/0xb4
[   10.493565] sp : ffffffc012a1b840
[   10.493863] x29: ffffffc012a1b840 x28: 0000000000000002
[   10.494335] x27: 0000000000000013 x26: ffffffc0112c52e8
[   10.494806] x25: ffffffc009410390 x24: 000000000000000a
[   10.495276] x23: ffffff8002c20c80 x22: ffffff8002c20ca8
[   10.495746] x21: 000000000000003d x20: ffffff8002c6c800
[   10.496217] x19: 0000000000000000 x18: 0000000000000020
[   10.496687] x17: 0000000000000000 x16: 0000000000000000
[   10.497158] x15: ffffff8006eb2d8a x14: ffffffffffffffff
[   10.497628] x13: ffffff8006eb2d88 x12: 0000000000000030
[   10.498099] x11: 0101010101010101 x10: 7f7f7f7f7f7f7f7f
[   10.498569] x9 : 626471716e626d48 x8 : 7f7f7f7f7f7f7f7f
[   10.499040] x7 : feff716475687163 x6 : 0000000000008080
[   10.499510] x5 : ffffffc0104f6960 x4 : ffffffc012a1b91c
[   10.499981] x3 : ffffffc0104f69d0 x2 : 0000000000000000
[   10.500451] x1 : 0000000000000000 x0 : ffffff8002c20c80
[   10.500922] Call trace:
[   10.501143]  dw_pcie_link_up+0x10/0x40
[   10.501476]  dw_pcie_other_conf_map_bus+0x34/0xb4
[   10.501890]  pci_generic_config_read+0x24/0xe0
[   10.502280]  dw_pcie_rd_other_conf+0x18/0x70
[   10.502660]  pci_bus_read_config_byte+0x60/0xa0
[   10.503061]  pci_read_config_byte+0x30/0x50
[   10.503432]  pci_assign_irq+0x38/0xd0
[   10.503754]  pci_device_probe+0x28/0x140
[   10.504102]  really_probe+0xf0/0x3d0
[   10.504424]  driver_probe_device+0x58/0xbc
[   10.504791]  device_driver_attach+0xc0/0xd0
[   10.505159]  __driver_attach+0x68/0xf4
[   10.505492]  bus_for_each_dev+0x60/0xa0
[   10.505837]  driver_attach+0x24/0x30
[   10.506158]  bus_add_driver+0x108/0x1fc
[   10.506502]  driver_register+0x78/0x130
[   10.506847]  __pci_register_driver+0x64/0x74
[   10.507236]  init_module+0x4c/0x1000 [r8125]
[   10.507618]  do_one_initcall+0x50/0x1c0
[   10.507964]  do_init_module+0x44/0x230
[   10.508298]  load_module+0x1e9c/0x25d0
[   10.508631]  __do_sys_init_module+0x194/0x270
[   10.509020]  __arm64_sys_init_module+0x1c/0x30
[   10.509413]  el0_svc_common.constprop.0+0x64/0x114
[   10.509837]  do_el0_svc+0x28/0x90
[   10.510139]  el0_svc+0x14/0x20
[   10.510416]  el0_sync_handler+0xa4/0x130
[   10.510760]  el0_sync+0x164/0x180
[   10.511065] Code: d503233f a9bf7bfd 910003fd f940cc02 (f9401042)
[   10.511605] ---[ end trace db3c3bd6b4d1de5a ]---
[   10.513922] Kernel panic - not syncing: Oops: Fatal exception
```
---
 drivers/pci/controller/dwc/pcie-dw-rockchip.c | 6 ++++++
 1 file changed, 6 insertions(+)

--- a/drivers/pci/controller/dwc/pcie-dw-rockchip.c
+++ b/drivers/pci/controller/dwc/pcie-dw-rockchip.c
@@ -854,6 +854,12 @@ static int rk_pcie_init_dma_trx(struct r
 
 	rk_pcie->dma_obj = rk_pcie_dma_obj_probe(rk_pcie->pci->dev);
 	if (IS_ERR(rk_pcie->dma_obj)) {
+		if (rk_pcie->mode == RK_PCIE_RC_TYPE) {
+			/* dma_obj is optional for RC mode */
+			rk_pcie->dma_obj = NULL;
+			return 0;
+		}
+
 		dev_err(rk_pcie->pci->dev, "failed to prepare dma object\n");
 		return -EINVAL;
 	} else if (rk_pcie->dma_obj) {
